compiler/src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -660,7 +665,11 @@ object SpaceEngine {
             else if tp.classSymbol == defn.TupleClass || tp.classSymbol == defn.NonEmptyTupleClass then
               List(child) // TupleN and TupleXXL classes are used for Tuple, but they aren't Tuple's children
             else if (child.is(Private) || child.is(Sealed)) && child.isOneOf(AbstractOrTrait) then getChildren(child)
-            else List(child)
+            else
+              // if a class contains a field of type Nothing,
+              // then it can be ignored in pattern matching, because it is impossible to obtain an instance of it
+              val noFieldsWithTypeNothing = child.typeRef.fields.forall(_.symbol.typeRef.classSymbol != defn.NothingClass)
+              if noFieldsWithTypeNothing then List(child) else Nil
```

Comment:
Why two changes?  Looks like this one might be the needed one.

Reply:
You mean why apply arguments to class type in function `getAppliedClass`?

Reply:
Yeah, but I think I understand now why you have to catch it twice.